docs/source/jit_language_reference_v2.rst

```diff
@@ -29,6 +29,213 @@
+                         FuncOrMethodBody
+    ParamAnnot := Identifier [ ":" TSType ] ","
+    ReturnAnnot := "->" TSType
+
+Note that using Python3 style, the type of ``self`` is automatically inferred and should not be annotated.
+
+* **Mypy style** annotates types as a comment right below the function/method declaration. In the My-Py style, since parameter names do not appear
+in the annotation, all parameters have to be annotated.
+
+::
+
+    MyPyAnnotation := "# type:" "(" ParamAnnot* ")" [ ReturnAnnot ]
+    ParamAnnot := TSType ","
+    ReturnAnnot := "->" TSType
+
+**Example 1**
+
+In this example, ``a`` is not annotated and assumes the default type of ``TensorType``, ``b`` is annotated as type ``int``, and the return type is not
+annotated and is automatically inferred as type ``TensorType`` (based on the type of the value being returned).
+
+::
+
+    import torch
+
+    def f(a, b: int):
+        return a+b
+
+    m = torch.jit.script(f)
+    print("TorchScript:", m(torch.ones([6]), 100))
+
+**Example 2**
+
+The following code snippet gives an example of using mypy style annotation. Note that parameters or return values must be annotated even if some of
+them assume the default type.
+
+::
+
+    import torch
+
+    def f(a, b):
+        # type: (torch.Tensor, int) → torch.Tensor
+        return a+b
+
+    m = torch.jit.script(f)
+    print("TorchScript:", m(torch.ones([6]), 100))
+
+
+Annotate variables and data attributes
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+In general, types of data attributes (including class and instance data attributes) and local variables can be automatically inferred from assignment statements.
+Sometimes, however, if a variable or attribute is associated with values of different types (e.g., as ``None`` or ``TensorType``), then they may need to be explicitly
+type annotated as a *wider* type such as ``Optional[int]`` or ``Any``.
+
+Local variables
+"""""""""""""""
+Local variables can be annotated according to Python3 typing module annotation rule, i.e.,
+
+::
+
+    LocalVarAnnotation := Identifier [":" TSType] "=" Expr
+
+In general, types of local variables can be automatically inferred. In some cases, however, programmers may need to annotate a multi-type for local variables
+that may be associated with different concrete types. Typical multi-types include ``Optional[T]`` and ``Any``.
+
+**Example**
+
+::
+
+    import torch
+
+    def f(a, setVal: bool):
+        value: Optional[torch.Tensor] = None
+        if setVal:
+            value = a
+        return value
+
+    ones = torch.ones([6])
+    m = torch.jit.script(f)
+    print("TorchScript:", m(ones, True), m(ones, False))
+
+Instance data attributes
+""""""""""""""""""""""""
+For ``ModuleType`` classes, instance data attributes can be annotated according to Python3 typing module annotation rules. Instance data attributes can be annotated (optionally) as final
+via ``torch.jit.Final``.
```

Comment:
Since `torch.jit.Final` is only kept for backward compatibility, we'd better use `Final` in the examples (and perhaps add a footnote that `torch.jit.Final` was introduced before `Final` was available and only kept for backward compatibility purposes)